crates/red_knot_test/README.md

```diff
@@ -0,0 +1,430 @@
+output of a test. Or sometimes (see “incremental tests” below) we will want to assert on diagnostics
+in a file, without impacting the contents of that file by changing a comment in it. In these cases,
+a Python fenced code block in a test could be followed by a fenced code block with language
+`output`; this contains the full diagnostic output for the preceding test file:
+
+````markdown
+# full output
+
+```py
+x = 1
+reveal_type(x)
+```
+
+This is just an example, not a proposal that red-knot would every actually output diagnostics in
+precisely this format:
+
+```output
+test.py, line 1, col 1: revealed type is 'Literal[1]'
+```
+````
+
+We will want to build tooling to automatically capture and update these “full diagnostic output”
+blocks, when tests are run in an update-output mode (probably specified by env var.)
+
+By default, an `output` block specifies diagnostic output for the file `<workspace-root>/test.py`.
+An `output` block can have a `path=` option, to explicitly specify the Python file for which it
+asserts diagnostic output, and a `stage=` option, to specify which stage of an incremental test it
+specifies diagnostic output at. (See “incremental tests” below.)
+
+It is an error for an `output` block to exist, if there is no `py` or `python` block in the same
+test for the same file path.
+
+### Incremental tests
+
+Some tests should validate incremental checking, by initially creating some files, checking them,
+and then modifying/adding/deleting files and checking again.
+
+We should add the capability to create an incremental test by using the `stage=` option on some
+fenced code blocks in the test:
+
+````markdown
+# Incremental
+
+## modify a file
+
+Initial version of `test.py` and `b.py`:
+
+```py
+from b import x
+reveal_type(x)
+```
+
+```py path=b.py
+x = 1
+```
+
+Initial expected output for `test.py`:
+
+```output
+/src/test.py, line 1, col 1: revealed type is 'Literal[1]'
+```
+
+Now in our first incremental stage, modify the contents of `b.py`:
+
+```py path=b.py stage=1
+# b.py
+x = 2
+```
+
+And this is our updated expected output for `test.py` at stage 1:
+
+```output stage=1
+/src/test.py, line 1, col 1: revealed type is 'Literal[2]'
+```
+
+(One reason to use full-diagnostic-output blocks in this test is that updating
+inline-comment diagnostic assertions for `test.py` would require specifying new
+contents for `test.py` in stage 1, which we don't want to do in this test.)
+````
+
+Any number of stages can be provided in an incremental test. If a stage re-specifies a filename that
+was specified in a previous stage (or the initial state), that file is modified. A new filename
+appearing for the first time in a new stage will create a new file. To delete a previously-created
+file, specify that file with the tag `delete` in its tag string (in this case, it is an error to
+provide non-empty contents.) Any previously-created files that are not re-specified in a later stage
```

Comment:
Closing parens should come before the period unless it's a full sentence inside the parens:

```suggestion
provide non-empty contents). Any previously-created files that are not re-specified in a later stage
```